### ISSUE TITLE
Grid Component

### ISSUE THREAD

**Issue by underlog** [telerik/kendo-angular2#1]
We should define the basic feature set / scope of the initial preview.

CC: @rkonstantinov @rusev

**Comment by HJoost** [comment]
I think a very must feature is to have checkboxes inside cells as representation of boolean values.

**Comment by rkonstantinov** [comment]
# Grid Component
The Grid is an Angular 2 component that displays data in a tabular format.

## Features
Planned features for the Beta release:

 - Three scrolling modes: Virtual, scrollable + pager and non-scrollable + pager
 - Databound columns as well as ability to customize the cell's appearance/show arbitrary content via templates (header, footer, data-cell)
 - Auto-generate columns definition based on the provided data
 - Detail Template
 - Toolbar template
 - Provide built-in UI for ordering (on single/multiple fields) the data by clicking on the column header
 - Provide built-in UI for navigating to a specific page of data and/or change the number of displayed records
 - Ability to lock/freeze columns - locked columns will be grouped at the most left side
 - Simple single row selection - user should be able to select/deselect row by clicking on it
 - Support for binding to RxJS Observables (via async pipe) and Arrays
 
The following features are unlikely to ship for the Beta release, but will follow in later releases:
 - Built-in UI for filtering
 - Displaying grouped data
 - More advanced selection + keyboard navigation
 - In-cell/In-line editing
 - Multiple column headers
 - Ability to re-order/re-size columns
 - Built-in UI for hide/show columns
 - PDF/Excel Export

## Sample Usage

The below example demonstrate two basic binding to local data scenarios - minimal configuration and a bit-more-advanced sorting/paging setup:
```
import 'core-js/es6';
import 'core-js/es7/reflect';
import 'zone.js/dist/zone';

import { Component } from '@angular/core';
import { bootstrap } from '@angular/platform-browser-dynamic';
import { GRID_DIRECTIVES } from '../src/grid.directives';
import { GridDataResult } from '../src/data.collection';
import { PageChangeEvent } from '../src/change-event-args.interface';

import { SortDescriptor } from '../src/sort-descriptor';
import { orderBy } from '../src/data/array.operators';

@Component({
  directives: [GRID_DIRECTIVES],
  selector: 'my-app',
  template: `
      <h1>Non-pagable local binding</h1>
      <kendo-grid [data]="grid1View"></kendo-grid>
      <h1>Local binding</h1>
      <kendo-grid
          [data]="grid2View"
          [pageSize]="pageSize"
          [skip]="skip"
          [sortable]="{ mode: 'multiple' }"
          [sort]="sort"
          [pageable]="true"
          (pageChange)="pageChange($event)"
          (sortChange)="sortChange($event)"
        >
        <kendo-grid-column field="ProductID" title="Product ID" width="120">
        </kendo-grid-column>
        <kendo-grid-column field="ProductName" title="Product Name">
        </kendo-grid-column>
        <kendo-grid-column field="UnitPrice" title="Unit Price" width="230">
        </kendo-grid-column>
        <kendo-grid-column field="Discontinued" width="120">
            <template kendoCellTemplate let-dataItem>
                <input type="checkbox" [checked]="dataItem.Discontinued" disabled/>
            </template>
        </kendo-grid-column>
      </kendo-grid>
  `
})
class AppComponent {

    private sort: SortDescriptor[] = [];
    private grid1View: any[];
    private grid2View: GridDataResult;

    private pageSize: number = 5;
    private skip: number = 0;

    constructor() {
        this.grid1View = products;

        this.loadProducts();
    }

    protected pageChange({ skip }: PageChangeEvent): void {
        this.skip = skip;
        this.loadProducts();
    }

    protected sortChange(sort: SortDescriptor[]): void {
        this.sort = sort;
        this.loadProducts();
    }

    private loadProducts(): void {
        this.grid2View = {
            data: orderBy(products, this.sort).slice(this.skip, this.skip + this.pageSize),
            total: products.length
        };
    }
}

const products = [{
    "ProductID": 1,
    "ProductName": "Chai",
    "UnitPrice": 18.0000,
    "Discontinued": false
  }, {
    "ProductID": 2,
    "ProductName": "Chang",
    "UnitPrice": 19.0000,
    "Discontinued": false
  }, /*..*/];

bootstrap(AppComponent);
```

**Comment by zhao125787935** [comment]
I am just wondering is there any possibility to support lazy loading( scrollable and no pager) in later releases?

**Comment by rkonstantinov** [comment]
We plan to support virtual scrolling (as mentioned in the features list above) with the initial release. This will allow for loading only portion of the data while user scrolls through the content.

**Comment by garethsuarez** [comment]
this looks great.  Most of the features I am waiting on are in the "upcoming release" category, but I would rather have something basic to add to the app and show off, than have to wait even longer for the first beta release with all the features I need.

any chance that the upcoming filter UI could be template driven? 

Also, locking columns, possible to choose to lock to the right as well as the left?

**Comment by duncan-mcleod** [comment]
Frozen rows would be a nice to have as well.

**Comment by masaanli** [comment]
Think 'Built-in UI for hide/show columns' is also an important one for initial release.

**Comment by rkonstantinov** [comment]
Thank you all for the suggestion.

@garethsuarez current implementation of the locked columns, allows locked columns to be placed only on the left side.

@duncan-mcleod I'm afraid that the frozen rows are not in our immediate plans.

**Comment by underlog** [comment]
The grid component is published at http://www.telerik.com/kendo-angular-ui/components/grid/

**Comment by clogeceyhan** [comment]
Can save the data as excel if we use grid component?

**Comment by underlog** [comment]
Hi @clogeceyhan,

This issue is a Roadmap item, which we already delivered - the features available are listed above. In case you need any kind of assistance, please follow the guidelines from [our web site](http://www.telerik.com/kendo-angular-ui/support/).
